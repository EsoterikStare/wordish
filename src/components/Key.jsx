import React from 'react';
import { shape, string } from 'prop-types';
import { Button, darken, Grid } from '@mui/material';

import { useAppState } from '../AppStateContext';

const Key = ({ label, guessedLetters }) => {
  const { dispatch } = useAppState();
  const dispatchAddAction = () => dispatch({ type: 'add', value: label });
  const handleKeyDown = (e) => {
    const actionKeys = [' ', 'Enter'];
    if (actionKeys.includes(e.key)) {
      dispatchAddAction();
    }
  };

  const keyStatus = guessedLetters[label.toLowerCase()];

  return (
    <Grid id="key-grid" item justifyContent="center">
      <Button
        aria-label={label}
        role="button"
        disableFocusRipple
        onClick={dispatchAddAction}
        onKeyDown={handleKeyDown}
        sx={[
          (theme) => ({
            backgroundColor: theme.palette.letterStatus[keyStatus] || '#555',
            color: 'white',
            fontSize: 20,
            height: 50,
            width: 35,
            minWidth: 35,
            userSelect: 'none'
          }),
          (theme) => ({
            '&:hover': {
              backgroundColor: theme.palette.letterStatus[keyStatus] || '#555'
            }
          }),
          (theme) => ({
            '&:active, &:focus, .Mui-focusVisible': {
              backgroundColor: theme.palette.letterStatus[keyStatus] || '#555'
            }
          })
        ]}
      >
        {label.toUpperCase()}
      </Button>
    </Grid>
  );
};

Key.propTypes = {
  label: string.isRequired,
  guessedLetters: shape({}).isRequired
};

export default Key;
