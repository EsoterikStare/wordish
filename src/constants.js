/* eslint-disable import/prefer-default-export */

export const letterStatus = {
  LOCATED: 'located',
  PRESENT: 'present',
  ABSENT: 'absent'
};
