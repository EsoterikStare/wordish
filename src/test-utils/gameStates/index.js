import loseGameStateJSON from './loseGameState.json';
import midGuessGameStateJSON from './midGuessGameState.json';
import newGameStateJSON from './newGameState.json';
import winGameStateJSON from './winGameState.json';

export const loseGameState = JSON.stringify(loseGameStateJSON);
export const midGuessGameState = JSON.stringify(midGuessGameStateJSON);
export const newGameState = JSON.stringify(newGameStateJSON);
export const winGameState = JSON.stringify(winGameStateJSON);
