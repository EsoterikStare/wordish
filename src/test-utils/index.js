export * from './gameStates';
export { default as setup } from './setup';
