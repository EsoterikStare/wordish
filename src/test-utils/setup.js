import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { node } from 'prop-types';
import React from 'react';

import { AppStateProvider } from '../AppStateContext';
import AppThemeProvider from '../AppThemeProvider';

const Wrapper = ({ children }) => (
  <AppStateProvider>
    <AppThemeProvider>{children}</AppThemeProvider>
  </AppStateProvider>
);
Wrapper.propTypes = {
  children: node.isRequired,
};
Wrapper.defaultProps = {};

const setup = jsx => {
  const { rerender, ...rest } = render(<Wrapper>{jsx}</Wrapper>);
  return {
    user: userEvent.setup(),
    rerender: newJsx => rerender(<Wrapper>{newJsx}</Wrapper>),
    ...rest,
  };
};

export default setup;
