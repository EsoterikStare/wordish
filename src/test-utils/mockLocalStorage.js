import { newGameState } from './gameStates';

const mockLocalStorage = (gameState = newGameState) => {
  const originalLocalStorage = { ...window.localStorage };
  const mockValues = {
    clear: jest.fn(),
    // Provide a known win state from localStorage mock that will be used by AppStateProvider.
    getItem: jest.fn(() => gameState),
    removeItem: jest.fn(),
    setItem: jest.fn(),
  };

  // Mock window.localStorage with assertable functions.
  Object.defineProperty(window, 'localStorage', {
    value: mockValues,
    writable: true,
  });

  const resetLocalStorage = () => {
    Object.defineProperty(window, 'localStorage', {
      value: originalLocalStorage,
    });
  };

  return {
    resetLocalStorage,
    ...mockValues,
  };
};

export default mockLocalStorage;
