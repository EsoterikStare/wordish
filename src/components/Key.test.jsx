import { screen } from '@testing-library/react';
import React from 'react';

import Key from './Key';

import { setup } from '../test-utils';
import { letterStatus } from '../constants';

describe('Key', () => {
  it('should be defined', () => {
    expect(Key).toBeDefined();
  });
  it('should render', () => {
    setup(<Key label="a" guessedLetters={{}} />);
    expect(screen.getByRole('button', { name: /a/i })).toBeInTheDocument();
  });
  it('should always render a capital letter', () => {
    setup(<Key label="a" guessedLetters={{}} />);
    expect(screen.getByRole('button', { name: 'A' })).toBeInTheDocument();
  });
  it('should look correct for a guessed value', () => {
    setup(<Key label="a" guessedLetters={{}} />);
    expect(screen.getByRole('button', { name: /a/i })).toMatchSnapshot();
  });
  it('should look correct for an absent value', () => {
    setup(<Key label="a" guessedLetters={{ a: letterStatus.ABSENT }} />);
    expect(screen.getByRole('button', { name: /a/i })).toMatchSnapshot();
  });
  it('should look correct for an present value', () => {
    setup(<Key label="a" guessedLetters={{ a: letterStatus.PRESENT }} />);
    expect(screen.getByRole('button', { name: /a/i })).toMatchSnapshot();
  });
  it('should look correct for an located value', () => {
    setup(<Key label="a" guessedLetters={{ a: letterStatus.LOCATED }} />);
    expect(screen.getByRole('button', { name: /a/i })).toMatchSnapshot();
  });
});
