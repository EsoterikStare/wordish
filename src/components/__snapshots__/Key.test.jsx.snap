// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Key should look correct for a guessed value 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-98k53l-MuiButtonBase-root-MuiButton-root"
  role="button"
  tabindex="0"
  type="button"
>
  A
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`Key should look correct for an absent value 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-wxkx1q-MuiButtonBase-root-MuiButton-root"
  role="button"
  tabindex="0"
  type="button"
>
  A
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`Key should look correct for an located value 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-qu45c3-MuiButtonBase-root-MuiButton-root"
  role="button"
  tabindex="0"
  type="button"
>
  A
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`Key should look correct for an present value 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-iep6mo-MuiButtonBase-root-MuiButton-root"
  role="button"
  tabindex="0"
  type="button"
>
  A
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
